File: clients/typescript/src/live/cache.ts
import type { ZbRecord } from "../realtime.js";

/** Framework-agnostic observable contract (matches the spec). */
export interface Observable<T> {
  subscribe(cb: () => void): () => void;
  get(): T;
  readonly version: number;
}

/**
 * A wrapped record that "looks exactly like the record it wraps": every
 * enumerable field of the backing object is exposed as a same-named getter, so
 * `live.title` reads through to the current backing data. Identity is stable —
 * updates patch the backing object IN PLACE and bump `version`; deletes set
 * `deleted = true`. Bind via the Observable contract.
 */
export class LiveRecord<T extends ZbRecord = ZbRecord> implements Observable<T> {
  deleted = false;
  private _version = 0;
  private backing: T;
  private readonly listeners = new Set<() => void>();
  private readonly proxied = new Set<string>();

  constructor(initial: T) {
    this.backing = initial;
    this.defineAccessors(Object.keys(initial));
  }

  get id(): string {
    return this.backing.id;
  }

  get version(): number {
    return this._version;
  }

  get(): T {
    return this.backing;
  }

  /** Returns the current data; alias kept for ergonomics. */
  get data(): T {
    return this.backing;
  }

  subscribe(cb: () => void): () => void {
    this.listeners.add(cb);
    return () => this.listeners.delete(cb);
  }

  /** Patch backing fields in place (used by the cache on update events). */
  patch(next: T): void {
    // Mutate the SAME backing object so external `.get()` references stay live.
    const target = this.backing as Record<string, unknown>;
    for (const k of Object.keys(target)) {
      if (!(k in next)) delete target[k];
    }
    Object.assign(target, next);
    this.defineAccessors(Object.keys(next));
    this.bump();
  }

  markDeleted(): void {
    this.deleted = true;
    this.bump();
  }

  private defineAccessors(keys: string[]): void {
    for (const key of keys) {
      if (key === "id" || key === "version" || key === "deleted" || this.proxied.has(key)) {
        continue;
      }
      this.proxied.add(key);
      Object.defineProperty(this, key, {
        configurable: true,
        enumerable: true,
        get: () => (this.backing as Record<string, unknown>)[key],
      });
    }
  }

  private bump(): void {
    this._version += 1;
    for (const cb of this.listeners) cb();
  }
}

interface Entry {
  record: LiveRecord;
  refs: number;
}

/**
 * Per-collection record cache keyed by record id. Hands out the SAME LiveRecord
 * for a given id so one event updates every view. Ref-counted: retained while
 * ≥1 live view/observer references it; evicted at zero.
 */
export class RecordCache {
  private readonly entries = new Map<string, Entry>();

  /** Retain (and create-or-merge) the record for `data.id`; +1 refcount. */
  retain<T extends ZbRecord>(data: T): LiveRecord<T> {
    const existing = this.entries.get(data.id);
    if (existing) {
      existing.refs += 1;
      // Merge fresher data without losing identity.
      existing.record.patch(data);
      return existing.record as LiveRecord<T>;
    }
    const record = new LiveRecord<T>(data);
    this.entries.set(data.id, { record, refs: 1 });
    return record;
  }

  get<T extends ZbRecord = ZbRecord>(id: string): LiveRecord<T> | undefined {
    return this.entries.get(id)?.record as LiveRecord<T> | undefined;
  }

  has(id: string): boolean {
    return this.entries.has(id);
  }

  release(id: string): void {
    const entry = this.entries.get(id);
    if (!entry) return;
    entry.refs -= 1;
    if (entry.refs <= 0) this.entries.delete(id);
  }

  applyUpdate<T extends ZbRecord>(data: T): void {
    this.entries.get(data.id)?.record.patch(data);
  }

  applyDelete(id: string): void {
    this.entries.get(id)?.record.markDeleted();
  }
}

File: clients/dart/lib/src/live/cache.dart
/// The observable contract, the live-record wrapper, and the ref-counted
/// per-collection record cache.
///
/// A behavioral port of `clients/typescript/src/live/cache.ts`, adapted to a
/// pure-Dart, framework-agnostic streams model (see
/// `docs/superpowers/specs/2026-07-08-dart-live-store-design.md`).
library;

import 'dart:async';

import '../records.dart';

/// Reserved fields owned by the wrapper; never patched from a server payload.
const _reservedKeys = {'id', 'version', 'deleted'};

/// Keys that could reassign a prototype / mutate intrinsics in JS. Harmless in
/// Dart, but dropped from patches for cross-SDK parity and defense-in-depth.
const _pollutingKeys = {'__proto__', 'constructor', 'prototype'};

bool _isReservedKey(String key) => _reservedKeys.contains(key);

bool _isSafePatchKey(String key) =>
    !_reservedKeys.contains(key) && !_pollutingKeys.contains(key);

/// Framework-agnostic observable contract (matches the spec).
///
/// [get] is the synchronous current snapshot; [version] a monotonic counter
/// bumped BEFORE each change notification; [changes] a broadcast stream that
/// emits one void event per mutation. `changes.listen(...)` is the TS
/// `subscribe(cb)`; cancelling the subscription is the returned unsubscribe.
abstract class Observable<T> {
  /// The current backing value.
  T get();

  /// A monotonic counter incremented on every mutation, before [changes] fires.
  int get version;

  /// A broadcast stream nudged once per mutation.
  Stream<void> get changes;
}

/// A wrapped record with a stable identity. Updates patch the backing record
/// IN PLACE and bump [version]; deletes set [deleted]. Bind via the
/// [Observable] contract; read fields through [get] (a [ZbRecord]) or the
/// convenience [operator []].
class LiveRecord implements Observable<ZbRecord> {
  bool _deleted = false;
  int _version = 0;
  final ZbRecord _backing;
  final StreamController<void> _changes = StreamController<void>.broadcast();

  LiveRecord(ZbRecord initial) : _backing = initial;

  /// The record id.
  String get id => _backing.id;

  /// `true` once a delete event has been applied.
  bool get deleted => _deleted;

  @override
  int get version => _version;

  @override
  ZbRecord get() => _backing;

  /// The current data; alias of [get] kept for ergonomics.
  ZbRecord get data => _backing;

  @override
  Stream<void> get changes => _changes.stream;

  /// Reads a raw backing field. Replaces TS's dynamic `live.title` getter,
  /// which Dart cannot synthesize at runtime.
  dynamic operator [](String key) => _backing.data[key];

  /// Patch backing fields in place (used by the cache on update events).
  ///
  /// Mutates the SAME backing map so external [get] references stay live. Only
  /// safe own keys are copied; reserved fields (`id`/`version`/`deleted`) are
  /// owned by the wrapper and left untouched, and `__proto__`/`constructor`/
  /// `prototype` are dropped.
  void patch(ZbRecord next) {
    final target = _backing.data;
    final safeKeys = next.data.keys.where(_isSafePatchKey).toList();
    final safeSet = safeKeys.toSet();
    // Drop keys that disappeared from the payload (but keep reserved fields).
    for (final k in target.keys.toList()) {
      if (!safeSet.contains(k) && !_isReservedKey(k)) {
        target.remove(k);
      }
    }
    for (final k in safeKeys) {
      target[k] = next.data[k];
    }
    _bump();
  }

  /// Flag this record deleted and notify.
  void markDeleted() {
    _deleted = true;
    _bump();
  }

  void _bump() {
    _version += 1;
    if (!_changes.isClosed) _changes.add(null);
  }

  /// Close the change stream. Called by [RecordCache] on eviction; not part of
  /// the public API.
  void dispose() {
    if (!_changes.isClosed) unawaited(_changes.close());
  }
}

class _Entry {
  final LiveRecord record;
  int refs;
  _Entry(this.record, this.refs);
}

/// Per-collection record cache keyed by record id. Hands out the SAME
/// [LiveRecord] for a given id so one event updates every view. Ref-counted:
/// retained while >= 1 live view/observer references it; evicted at zero.
class RecordCache {
  final Map<String, _Entry> _entries = {};

  /// Retain (and create-or-merge) the record for `data.id`; +1 refcount.
  LiveRecord retain(ZbRecord data) {
    final existing = _entries[data.id];
    if (existing != null) {
      existing.refs += 1;
      // Merge fresher data without losing identity.
      existing.record.patch(data);
      return existing.record;
    }
    final record = LiveRecord(data);
    _entries[data.id] = _Entry(record, 1);
    return record;
  }

  /// The cached record for [id], or null.
  LiveRecord? get(String id) => _entries[id]?.record;

  /// Whether [id] is currently retained.
  bool has(String id) => _entries.containsKey(id);

  /// Release one ref for [id]; evict and dispose the record at zero.
  void release(String id) {
    final entry = _entries[id];
    if (entry == null) return;
    entry.refs -= 1;
    if (entry.refs <= 0) {
      _entries.remove(id);
      entry.record.dispose();
    }
  }

  /// Apply an update event to the cached record for `data.id`, if present.
  void applyUpdate(ZbRecord data) => _entries[data.id]?.record.patch(data);

  /// Apply a delete event to the cached record for [id], if present.
  void applyDelete(String id) => _entries[id]?.record.markDeleted();
}
